### PR TITLE
NF: Applying Decorators in Module (Align)

### PR DESCRIPTION
**Pull request by deka27**
## This PR is to apply keyword decorator to functions in align module.

This PR is dependent on the main PR, which should be merged first.

## Main PR:
- #3239

### CODE REVIEW

File: dipy/align/cpd.py [hunk 1]
```diff
@@ -36,11 +39,11 @@ def low_rank_eigen(G, num_eig):
     Enables lower dimensional solving.
 
     """
-    S, Q = np.linalg.eigh(G)
+    (S, Q) = np.linalg.eigh(G)
     eig_indices = list(np.argsort(np.abs(S))[::-1][:num_eig])
-    Q = Q[:, eig_indices]  # eigenvectors
```
Comment (itellaetxe) [inline dipy/align/cpd.py:41]: Why did you remove the inline comments?

File: dipy/align/cpd.py [hunk 2]
```diff
@@ -89,17 +93,12 @@ def lowrankQS(G, beta, num_eig, eig_fgt=False):
...
 
     """
-    # if we do not use FGT we construct affinity matrix G and find the
-    # first eigenvectors/values directly
-
     if eig_fgt is False:
-        S, Q = np.linalg.eigh(G)
+        (S, Q) = np.linalg.eigh(G)
         eig_indices = list(np.argsort(np.abs(S))[::-1][:num_eig])
-        Q = Q[:, eig_indices]  # eigenvectors
```
Comment (itellaetxe) [inline dipy/align/cpd.py:98]: Same, inline comments are not there anymore

File: dipy/align/cpd.py [hunk 3]
```diff
@@ -325,29 +315,21 @@ def update_transform(self):
             )
             B = self.PX - np.dot(np.diag(self.P1), self.Y)
             self.W = np.linalg.solve(A, B)
-
         elif self.low_rank is True:
-            # Matlab code equivalent can be found here:
```
Comment (itellaetxe) [inline dipy/align/cpd.py:330]: Why wa this comment removed?

File: dipy/align/cpd.py [hunk 4]
```diff
@@ -397,27 +378,17 @@ def update_variance(self):
 
         """
         qprev = self.sigma2
-
-        # The original CPD paper does not explicitly calculate the objective
```
Comment (itellaetxe) [inline dipy/align/cpd.py:401]: Why did you remove this comment?

File: dipy/align/cpd.py [hunk 5]
```diff
@@ -397,27 +378,17 @@ def update_variance(self):
...
             np.transpose(self.P1), np.sum(np.multiply(self.TY, self.TY), axis=1)
         )
         trPXY = np.sum(np.multiply(self.TY, self.PX))
-
         self.sigma2 = (xPx - 2 * trPXY + yPy) / (self.Np * self.D)
-
         if self.sigma2 <= 0:
             self.sigma2 = self.tolerance / 10
-
-        # Here we use the difference between the current and previous
```
Comment (itellaetxe) [inline dipy/align/cpd.py:419]: Same, comment removed.

File: dipy/align/cpd.py [hunk 6]
```diff
@@ -452,10 +422,8 @@ def expectation(self):
             * self.M
             / self.N
         )
-
-        den = np.sum(P, axis=0, keepdims=True)  # (1, N)
+        den = np.sum(P, axis=0, keepdims=True)
```
Comment (itellaetxe) [inline dipy/align/cpd.py:425]: Comment removed.

File: dipy/align/reslice.py [hunk 7]
```diff
@@ -71,11 +73,6 @@ def reslice(
 
     """
     num_processes = determine_num_processes(num_processes)
-
-    # We are suppressing warnings emitted by scipy >= 0.18,
```
Comment (itellaetxe) [inline dipy/align/reslice.py:75]: Why did you remove the comment?

File: dipy/align/_public.py [hunk 8]
```diff
@@ -504,55 +493,42 @@ def affine_registration(
...
     affreg = AffineRegistration(
         metric=use_metric, level_iters=level_iters, sigmas=sigmas, factors=factors
     )
-
-    # Convert pipeline to sanitized list of str
     pipeline = list(pipeline)
     for fi, func in enumerate(pipeline):
         if callable(func):
             for key, val in _METHOD_DICT.items():
-                if func is val[0]:  # if they passed the callable equiv.
```
Comment (itellaetxe) [inline dipy/align/_public.py:521]: Removed comment

File: dipy/align/_public.py [hunk 9]
```diff
@@ -504,55 +493,42 @@ def affine_registration(
...
-                f"pipeline[{fi}] must be one of " f"{list(_METHOD_DICT)}, got {func!r}"
+                f"pipeline[{fi}] must be one of {list(_METHOD_DICT)}, got {func!r}"
             )
-
     if pipeline == ["center_of_mass"] and ret_metric:
         raise ValueError(
             "center of mass registration cannot return any quality metric."
         )
-
-    # Go through the selected transformation:
```
Comment (itellaetxe) [inline dipy/align/_public.py:534]: Removed comment

File: dipy/align/_public.py [hunk 10]
```diff
@@ -504,55 +493,42 @@ def affine_registration(
...
     for func in pipeline:
         if func == "center_of_mass":
             if starting_affine is not None and starting_was_supplied:
                 wm = "starting_affine overwritten by center_of_mass transform"
                 warn(wm, UserWarning, stacklevel=2)
-
-            # multiply images by masks for transform_centers_of_mass
```
Comment (itellaetxe) [inline dipy/align/_public.py:541]: Removed comment

File: dipy/align/_public.py [hunk 11]
```diff
@@ -565,51 +541,37 @@ def affine_registration(
                 moving_mask=moving_mask,
             )
             starting_affine = xform.affine
-
-    # Copy the final affine into a final variable
```
Comment (itellaetxe) [inline dipy/align/_public.py:569]: Removed comment

File: dipy/align/_public.py [hunk 12]
```diff
@@ -565,51 +541,37 @@ def affine_registration(
...
     final_affine = starting_affine.copy()
-
-    # After doing all that, resample once at the end:
```
Comment (itellaetxe) [inline dipy/align/_public.py:572]: Removed comment

File: dipy/align/_public.py [hunk 13]
```diff
@@ -565,51 +541,37 @@ def affine_registration(
...
     affine_map = AffineMap(
         final_affine, static.shape, static_affine, moving.shape, moving_affine
     )
-
     resampled = affine_map.transform(moving)
-
-    # Return the optimization metric only if requested
```
Comment (itellaetxe) [inline dipy/align/_public.py:579]: Removed comment

File: dipy/align/_public.py [hunk 14]
```diff
@@ -565,51 +541,37 @@ def affine_registration(
...
 rigid_scaling = partial(affine_registration, pipeline=["rigid_scaling"])
 rigid_scaling.__doc__ = (
     "Implements a rigid scaling transform. Based on `affine_registration()`."
 )
-
 affine = partial(affine_registration, pipeline=["affine"])
 affine.__doc__ = "Implements an affine transform. Based on `affine_registration()`."
-
-
-_METHOD_DICT = {  # mapping from str key -> (callable, class) tuple
```
Comment (itellaetxe) [inline dipy/align/_public.py:612]: Removed comment

File: dipy/align/_public.py [hunk 15]
```diff
@@ -666,23 +628,21 @@ def register_series(
...
                 "The reference image should be a single volume",
                 " or the index of one or more volumes",
             )
-
     xformed = np.zeros(series.shape)
     affines = np.zeros((4, 4, series.shape[-1]))
     for ii in range(series.shape[-1]):
         this_moving = series[..., ii]
         if isinstance(ref_as_idx, numbers.Number) and ii == ref_as_idx:
-            # This is the reference! No need to move and the xform is I(4):
```
Comment (itellaetxe) [inline dipy/align/_public.py:681]: Removed comment

File: dipy/align/_public.py [hunk 16]
```diff
@@ -738,59 +698,48 @@ def register_dwi_series(
...
     """
     pipeline = pipeline or ["center_of_mass", "translation", "rigid", "affine"]
-
-    data, affine = read_img_arr_or_path(data, affine=affine)
+    (data, affine) = read_img_arr_or_path(data, affine=affine)
     if isinstance(gtab, collections.abc.Sequence):
         gtab = dpg.gradient_table(*gtab)
-
     if np.sum(gtab.b0s_mask) > 1:
-        # First, register the b0s into one image and average:
```
Comment (itellaetxe) [inline dipy/align/_public.py:747]: Removed comment

File: dipy/align/_public.py [hunk 17]
```diff
@@ -738,59 +698,48 @@ def register_dwi_series(
...
         b0_img = nib.Nifti1Image(data[..., gtab.b0s_mask], affine)
-        trans_b0, b0_affines = register_series(
+        (trans_b0, b0_affines) = register_series(
             b0_img, ref=b0_ref, pipeline=pipeline, static_mask=static_mask
         )
         ref_data = np.mean(trans_b0, -1, keepdims=True)
     else:
-        # There's only one b0 and we register everything to it
```
Comment (itellaetxe) [inline dipy/align/_public.py:754]: Removed comment

File: dipy/align/_public.py [hunk 18]
```diff
@@ -738,59 +698,48 @@ def register_dwi_series(
...
         trans_b0 = ref_data = data[..., gtab.b0s_mask]
         b0_affines = np.eye(4)[..., np.newaxis]
-
-    # Construct a series out of the DWI and the registered mean B0:
```
Comment (itellaetxe) [inline dipy/align/_public.py:758]: Removed comment

File: dipy/align/_public.py [hunk 19]
```diff
@@ -738,59 +698,48 @@ def register_dwi_series(
...
     moving_data = data[..., ~gtab.b0s_mask]
     series_arr = np.concatenate([ref_data, moving_data], -1)
     series = nib.Nifti1Image(series_arr, affine)
-
-    xformed, affines = register_series(
+    (xformed, affines) = register_series(
         series, ref=0, pipeline=pipeline, static_mask=static_mask
     )
-    # Cut out the part pertaining to that first volume:
```
Comment (itellaetxe) [inline dipy/align/_public.py:766]: Removed comment

File: dipy/align/_public.py [hunk 20]
```diff
@@ -816,21 +765,17 @@ def streamline_registration(moving, static, n_points=100, native_resampled=False
         The affine transformation that takes us from 'moving' to 'static'
 
     """
-    # Load the streamlines, if you were given a file-name
```
Comment (itellaetxe) [inline dipy/align/_public.py:819]: Removed comment

**Review by itellaetxe — APPROVED** [review]
Looking good in general. However, I think you accidentally deleted some useful comments. Perhaps `ruff` messed this up? idk. I marked them, I think I did not miss any.

Other than that, I would approve this.

File: dipy/align/cpd.py (hunk 1 above)
Comment (deka27) [reply dipy/align/cpd.py:41]: I will get this fixed.

**Comment by skoudoro** [comment]
Can you rebase and fix the format since #3239 is merged ? Thanks!

File: dipy/align/reslice.py (hunk 7 above)
Comment (skoudoro) [reply dipy/align/reslice.py:75]: Why did you remove the comment?

File: dipy/align/cpd.py [hunk 21]
```diff
@@ -196,7 +201,6 @@ def __init__(
         max_iterations=None,
         tolerance=None,
         w=None,
-        *args,
```
Comment (skoudoro) [inline dipy/align/cpd.py:199]: Why do you remove *args ?

**Review by skoudoro — COMMENTED** [review]
Thank you for the update @deka27.

See below some comments before merging

File: dipy/align/cpd.py (hunk 21 above)
Comment (deka27) [reply dipy/align/cpd.py:199]: let me fix this, i think it happened while i was merging conflicts for rebasing.

File: dipy/align/cpd.py [hunk 22]
```diff
@@ -184,10 +187,13 @@ class DeformableRegistration:
         Number of eigenvectors to use in lowrank calculation.
     """
 
+    @warning_for_keywords()
     def __init__(
         self,
         X,
         Y,
+        args=(),
+        *,
```
Comment (skoudoro) [inline dipy/align/cpd.py:196]: it does not look correct. read and try this: https://stackoverflow.com/questions/64423083/why-cant-both-args-and-keyword-only-arguments-be-mixed-with-args-and-kwargs

**Review by skoudoro — COMMENTED** [review]
Hi @deka27,

see my comment below

File: dipy/align/cpd.py (hunk 22 above)
Comment (deka27) [reply dipy/align/cpd.py]: got it. corrected.

Comment (skoudoro) [reply dipy/align/cpd.py]: I think your decorator does not work with that. I tried:

```python
@warning_for_keywords(from_version="1.0.0", until_version="10.0.0")
def func_with_kwonly_args(a, b, *args, c=3):
      return a + b + c


func_with_kwonly_args(1, 2, 4)
```

and it crashes. Can you look into it and add a test in a new PR.  Thanks

Comment (skoudoro) [reply dipy/align/cpd.py]: Ah ok, after multiple tests, you do not need the decorator when you have `
*args`. Which is a good news !

File: dipy/align/cpd.py [hunk 23]
```diff
@@ -184,10 +187,12 @@ class DeformableRegistration:
         Number of eigenvectors to use in lowrank calculation.
     """
 
+    @warning_for_keywords()
```
Comment (skoudoro) [inline dipy/align/cpd.py:190]: decorator to remove here

**Review by skoudoro — COMMENTED** [review]
Hi @deka27, see below my comments

**Review by skoudoro — APPROVED** [review]
All good, Thanks @deka27, merging
